### PR TITLE
caching and async -> speeed

### PR DESCRIPTION
**Pull request by qttylog**
lets have some fun fun fun fun fun fun fun

### CODE REVIEW

File: menstruation/config.py [hunk 1]
```diff
@@ -80,6 +80,20 @@ def remove_user(self, user_id: int) -> int:
...
 
+try:
+    workers = int(os.environ["MENSTRUATION_WORKERS"])
+    if not workers:
+        raise KeyError
+except (KeyError, ValueError):
+    workers = 8
+
+try:
+    retries_api_failure = int(os.environ["MENSTRUATION_RETRIES_ON_API_FAILURE"])
```
Comment (kmein) [inline menstruation/config.py]: maybe use a shorter name here: MENSTRUATION_RETRIES

File: Dockerfile [hunk 2]
```diff
@@ -4,7 +4,9 @@ ADD . /app/
 
 ENV TZ=Europe/Berlin \
     MENSTRUATION_TOKEN='TOKEN' \
-    MENSTRUATION_ENDPOINT=http://127.0.0.1:8000
+    MENSTRUATION_ENDPOINT=http://127.0.0.1:8000 \
+    MENSTRUATION_MODERATORS="" \
```
Comment (kmein) [inline Dockerfile:8]: isn't that the default?

File: Dockerfile [hunk 3]
```diff
@@ -4,7 +4,9 @@ ADD . /app/
...
+    MENSTRUATION_WORKERS=8
```
Comment (kmein) [inline Dockerfile:9]: same as above

File: menstruation/handlers.py [hunk 4]
```diff
@@ -254,15 +277,40 @@ def unsubscribe_handler(update: Update, context: CallbackContext):
...
+            f"*User DB*\n"
+            f"Registered: {len(user_db.users())}\n"
+            f"Subscribed: {len(list(user for user in user_db.users() if user_db.is_subscriber(user)))}\n\n"
+            f"*Config*\n"
+            f"workers: {config.workers}\n"
+            f"moderators: {', '.join(config.moderators)}\n"
+            f"notification time: {config.notification_time.strftime('%H:%M')}\n"
+            f"retries on api failure:{config.retries_api_failure}\n"
+            f"debug: {config.debug}\n"
+            f"logging level: {logging.getLogger().getEffectiveLevel()}\n\n"
```
Comment (kmein) [inline menstruation/handlers.py]: Capitalize

File: menstruation/jobs.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+from time import sleep
+
+from emoji import emojize
+from telegram.error import Unauthorized
+
+from menstruation import config
+from telegram.ext import CallbackContext, JobQueue, Job
+
+from menstruation.handlers import send_menu
+from menstruation.query import Query
```
Comment (kmein) [inline menstruation/jobs.py]: Sort

File: menstruation/jobs.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+
+def add_subscriber(user_id: str):
+    logging.debug(f"Added subscriber: {user_id}")
+    job_queue.run_daily(
+        notify_subscriber,
+        config.notification_time,
+        days=(0, 1, 2, 3, 4),
+        name=user_id,
+    )
```
Comment (kmein) [inline menstruation/jobs.py:35]: So we're back to this again? :sweat_smile:

File: menstruation/jobs.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    logging.debug(f"Removed subscriber: {user_id}")
+    for job in job_queue.get_jobs_by_name(user_id):
+        logging.debug(f"Job: {job.name}, {job.enabled}, {job.removed}")
+        job.schedule_removal()
+
+
+def notify_subscriber(context: CallbackContext):
+    user_id = context.job.name
+    if not user_db.is_subscriber(user_id):
+        logging.error(f"{user_id} is no subscriber, but had an subscription job")
```
Comment (kmein) [inline menstruation/jobs.py]: Grammar: had a subscription job.

File: menstruation/jobs.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        config.notification_time,
+        days=(0, 1, 2, 3, 4),
+        name=user_id,
+    )
+
+
+def remove_subscriber(user_id: str):
+    logging.debug(f"Removed subscriber: {user_id}")
+    for job in job_queue.get_jobs_by_name(user_id):
+        logging.debug(f"Job: {job.name}, {job.enabled}, {job.removed}")
```
Comment (kmein) [inline menstruation/jobs.py]: boolean blindness: better log `enabled: {job.enabled}, removed: {job.removed}`

File: menstruation/jobs.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    for retries in range(config.retries_api_failure):
+        try:
+            send_menu(context.bot, user_id, Query.from_text(filter_text))
+        except TypeError:
+            logging.exception(f"{user_id} has no mensa selected")
+        except JSONDecodeError:
+            logging.debug(
+                f"JSONDecodeError: Try number {retries + 1} / {config.retries_api_failure}"
+            )
+            sleep(randint(100, (200 + (200 * users_sum) / 100)))
```
Comment (kmein) [inline menstruation/jobs.py]: maybe factor this out or add a comment what this does, e.g. what the minimum/maximum duration of sleep is

also: 200 * x / 100 = 2 * x

File: menstruation/jobs.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    job_queue.run_once(startup_message, 0)
+
+    for user_id in user_db.users():
+        if user_db.is_subscriber(user_id):
+            add_subscriber(str(user_id))
+    job_queue.start()
+
+
+def show_job_queue() -> str:
+    text = ""
```
Comment (kmein) [inline menstruation/jobs.py]: `text = "\n".join("f"{job.name}: enabled: {job.enabled}, removed: {job.removed}" for job in job_queue.jobs())`

**Review by kmein — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (qttylog) [reply Dockerfile:8]: Yeah but this is okay and intended. In the container overview, everyone can see that there is an important setting that can be made here.

File: menstruation/jobs.py (hunk 6 above)
Comment (qttylog) [reply menstruation/jobs.py:35]: Haha yeah.
I have brought your old idea back to life, so we leave open the option to let people set their own times.

File: menstruation/jobs.py (hunk 8 above)
Comment (qttylog) [reply menstruation/jobs.py]: That should not be there xD

File: menstruation/jobs.py (hunk 9 above)
Comment (qttylog) [reply menstruation/jobs.py]: The brackets were even wrong. Well seen

**Review by kmein — APPROVED** [review]

**Pull request merged**